### PR TITLE
Implement domain removal in mod_roster

### PR DESCRIPTION
**Pull request by Premwoik**
This PR implements domain removal in mod_roster (only rdbms).

Proposed changes include:
* extracting some functions to allow reuse,
* handling domain removal in mod_roster_rdbms,
* implementation of domain_removal hook in the mod_roster,
* adding a test case to the domain_removal_SUITE.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/75012) / small_tests / 210cda0850ccabeb8a3ccc9f45314e7e5299169c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3291/75012/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3291/75012/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-09-22_12.58.59/lib.mongooseim.logs/run.2021-09-22_12.58.59/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/75043) / small_tests / dae20e05e168fde260911fa848a97b76910ab4f7
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3291/75043/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3291/75043/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-09-22_13.29.41/lib.mongooseim.logs/run.2021-09-22_13.29.41/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/75048) / small_tests / b9df2469cee48898169e2b1653484ba8883ce5ec
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3291/75048/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3291/75048/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2021-09-22_13.30.44/lib.mongooseim.logs/run.2021-09-22_13.30.45/suite.log.html)

File: big_tests/tests/domain_removal_SUITE.erl [hunk 1]
```diff
@@ -197,6 +202,33 @@ private_removal(Config) ->
...
       end).
 
+bobs_default_groups() -> [<<"friends">>].
+
+bobs_default_name() -> <<"Bobby">>.
+
+roster_removal(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}, {bob, 1}], fun(Alice, Bob) ->
+        %% add contact
+        Stanza = escalus_stanza:roster_add_contact(Bob, bobs_default_groups(),
```
Comment (vkatsuba) [inline big_tests/tests/domain_removal_SUITE.erl:212]: Why here is used `bobs_default_groups/0` and `bobs_default_name/0`? Maybe make sense change to:
```erlang
Stanza = escalus_stanza:roster_add_contact(Bob, [<<"friends">>], <<"Bobby">>),
```

File: src/mod_roster_mnesia.erl [hunk 2]
```diff
@@ -108,6 +109,13 @@ remove_user_t(_HostType, LUser, LServer) ->
                   mnesia:index_read(roster, US, #roster.us)),
     ok.
 
+%% There is no optimized way to remove a domain.
+%% We expect, that domain removal process would call remove_user instead
+%% for each user.
+-spec remove_domain_t(mongooseim:host_type(), jid:lserver()) -> ok.
+remove_domain_t(_HostType, _LServer) ->
```
Comment (vkatsuba) [inline src/mod_roster_mnesia.erl:116]: Maybe better argument name will be `_Domain`. Then you can update `spec` as:
```erlang
-spec remove_domain_t(HostType :: mongooseim:host_type(), Domain :: jid:lserver()) -> ok.
```

File: src/mod_roster_rdbms.erl [hunk 3]
```diff
@@ -114,6 +115,13 @@ remove_user_t(HostType, LUser, LServer) ->
     mongoose_rdbms:execute_successfully(HostType, roster_group_delete, [LServer, LUser]),
     ok.
 
+-spec remove_domain_t(mongooseim:host_type(), jid:lserver()) -> ok.
+remove_domain_t(HostType, LServer) ->
```
Comment (vkatsuba) [inline src/mod_roster_rdbms.erl:119]: Maybe better argument name will be `_Domain`. Then you can update `spec` as:
```erlang
-spec remove_domain_t(HostType :: mongooseim:host_type(), Domain :: jid:lserver()) -> ok.
```

File: src/mod_roster_riak.erl [hunk 4]
```diff
@@ -125,6 +126,13 @@ remove_user_t(HostType, LUser, LServer) ->
     mongoose_riak:delete(?ROSTER_BUCKET(HostType, LServer), LUser),
     ok.
 
+%% There is no optimized way to remove a domain.
+%% We expect, that domain removal process would call remove_user instead
+%% for each user.
+-spec remove_domain_t(mongooseim:host_type(), jid:lserver()) -> ok.
```
Comment (vkatsuba) [inline src/mod_roster_riak.erl:132]: Maybe better argument name will be `_Domain`. Then you can update `spec` as:
```erlang
-spec remove_domain_t(HostType :: mongooseim:host_type(), Domain :: jid:lserver()) -> ok.
```

File: big_tests/tests/domain_removal_SUITE.erl [hunk 5]
```diff
@@ -197,6 +202,33 @@ private_removal(Config) ->
         ?assert_equal_extra(<<>>, Val2, #{stanza => Res2})
       end).
 
+bobs_default_groups() -> [<<"friends">>].
+
+bobs_default_name() -> <<"Bobby">>.
+
+roster_removal(Config) ->
```
Comment (vkatsuba) [inline big_tests/tests/domain_removal_SUITE.erl:209]: Can you provide tests case where tables `rosterusers`/`rostergroups`/`roster_version` will checking? Eg: check count of `rosterusers` before and after remove domain?

File: src/mod_roster_mnesia.erl (hunk 2 above)
Comment (chrzaszcz) [reply src/mod_roster_mnesia.erl:116]: Argument name can be changed to `Domain`, but I don't think there is a need to repeat variable names in such a simple spec.

File: big_tests/tests/domain_removal_SUITE.erl (hunk 5 above)
Comment (chrzaszcz) [reply big_tests/tests/domain_removal_SUITE.erl:209]: I think we could check it in the same test case... just add calls to a function similar to `select_affs_by_room_id`.

**Review by chrzaszcz — COMMENTED** [review]
Code looks good to me. I don't have any extra comments, I responded to the existing ones.

File: big_tests/tests/domain_removal_SUITE.erl [hunk 6]
```diff
@@ -197,6 +208,48 @@ private_removal(Config) ->
...
+        Received3 = escalus:send_iq_and_wait_for_result(Alice, escalus_stanza:roster_get()),
+        escalus:assert(is_roster_result, Received3),
+        escalus:assert(count_roster_items, [0], Received3),
+
+        {selected, []} = select_rosterusers(host_type(), domain()),
+        {selected, []} = select_rostergroups(host_type(), domain()),
+        {selected, []} = select_roster_version(host_type(), domain())
+        end).
+
+select_rosterusers(HostType, Domain) ->
```
Comment (vkatsuba) [inline big_tests/tests/domain_removal_SUITE.erl:241]: I suppose this function can be simplified like(name can be used other):

```erlang
select_from_roster(Table) ->
    Query = "SELECT * FROM " ++ Table ++ " WHERE server='" ++ binary_to_list(domain()) ++ "'",
    {selected, Res} = rpc(mim(), mongoose_rdbms, sql_query, [host_type(), Query]),
    Res.
```
using:
```erlang
[] = select_from_roster("rosterusers")
[_] = select_from_roster("rostergroups")
[] = select_from_roster("roster_version")
```

**Review by vkatsuba — APPROVED** [review]
LGTM. Posted small comment, but since comment is related to tests, this can be done in any time.

**Pull request merged**